dynamiqs/qarray.py

```diff
@@ -0,0 +1,554 @@
+
+    def wrapper(self: QArray, *args, **kwargs) -> callable:
+        return self.__class__(method(self, *args, **kwargs), dims=self.dims)
+
+    return wrapper
+
+
+class QArray(eqx.Module):
+    r"""Quantum array object. DenseQArray is a wrapper around JAX arrays. It offers
+    convenience methods to handle more easily quantum states
+    (bras, kets and density matrices) and quantum operators.
+    If you come from QuTiP, this is the equivalent of the Qobj class.
+    """
+
+    dims: tuple[int, ...]
+
+    @property
+    @abstractmethod
+    def dtype(self) -> jnp.dtype:
+        """Data type of the quantum state.
+
+        Returns:
+             The data type of the quantum state.
+        """
+
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, ...]:
+        """Shape of the quantum state.
+
+        Returns:
+            The shape of the quantum state.
+        """
+
+    @property
+    def ndim(self) -> int:
+        """Number of dimensions of the quantum state.
+
+        Returns:
+            The number of dimensions of the quantum state.
+        """
+        return len(self.dims)
+
+    @property
+    @abstractmethod
+    def I(self) -> QArray:  # noqa: E743
+        """Identity operator compatible with the quantum state.
+
+        Returns:
+            The identity operator.
+        """
+
+    @property
+    @abstractmethod
+    def mT(self) -> QArray:
+        """Transpose of the quantum state.
+
+        Returns:
+            The transpose of the quantum state.
+        """
+
+    # methods
+    @abstractmethod
+    def conj(self) -> QArray:
+        """Conjugate of the quantum state.
+
+        Returns:
+            The conjugate of the quantum state.
+        """
+
+    @abstractmethod
+    def norm(self) -> Array:
+        """Norm of the quantum state.
+
+        Returns:
+            The norm of the quantum state.
+        """
+
+    @abstractmethod
+    def unit(self) -> QArray:
+        """Normalize the quantum state.
+
+        Returns:
+            The normalized quantum state.
+        """
+
+    @abstractmethod
+    def diag(self) -> Array:
+        """Diagonal of the quantum state.
+
+        Returns:
+            The diagonal of the quantum state.
+        """
+
+    @abstractmethod
+    def trace(self) -> Array:
+        """Trace of the quantum state.
+
+        Returns:
+            The trace of the quantum state.
+        """
+
+    @abstractmethod
+    def reshape(self, *new_shape: int) -> QArray:
+        """Reshape the quantum state.
+
+        Args:
+            new_shape: New shape of the quantum state.
+
+        Returns:
+            The reshaped quantum state.
+        """
+
+    @abstractmethod
+    def broadcast_to(self, *new_shape: int) -> QArray:
+        """Broadcast the quantum state.
+
+        Returns:
+            The broadcast quantum state.
+        """
+
+    @abstractmethod
+    def ptrace(self, keep_dims: tuple[int, ...]) -> QArray:
```

Comment:
```suggestion
    def ptrace(self, keep: tuple[int, ...]) -> QArray:
```
(renamed in the latest iteration of the RFC)